model_card_toolkit/base_model_card_field.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""BaseModelCardField.
+
+This class serves as a basic shared API between all Model Card data classes (
+see model_card.py).
+"""
+
+import abc
+import json as json_lib
+from typing import Any, Dict, List, Text, Union
+import dataclasses
+
+from google.protobuf import descriptor
+from google.protobuf import message
+
+MessageFieldType = Union["BaseModelCardField", Dict[Text, Any]]
+PrimitiveFieldType = Union[float, int, Text]
+
+
+class BaseModelCardField(abc.ABC):
+  """Model card field base class.
+
+  This is an abstract class. All the model card fields should inherit this class
+  and override the _proto_type property to the corresponding proto type. This
+  abstract class provides methods `copy_from_proto`, `merge_from_proto` and
+  `to_proto` to convert the class from and to proto. The child class does not
+  need to override this unless it needs some special process.
+  """
+
+  @property
+  @abc.abstractmethod
+  def _proto_type(self):
+    """The proto type. Child class should overwrite this."""
+
+  def is_empty(self) -> bool:
+    for field_val in self.__dict__.values():
+      if isinstance(field_val, BaseModelCardField):
+        return field_val.is_empty()
+      elif field_val:
+        return False
+    return True
+
+  def to_proto(self) -> message.Message:
+    """Convert this class object to the proto."""
+
+    def _is_empty_field(field_val: Any):
+      return not field_val or (isinstance(field_val, BaseModelCardField) and
+                               field_val.is_empty())
+
+    def _field_to_proto(field_val: MessageFieldType,
+                        field_name: Text) -> message.Message:
+      # Fields can be BaseModelCardField or Dict.
+      # Only BaseModelCardField can be converted to Proto.
+      # So if field is Dict, we need to convert it to BaseModelCardField first.
+      if isinstance(field_val, dict):
+        # Get field's class
+        if isinstance(getattr(self, field_name), list):
+          field_class = getattr(self, "__annotations__")[field_name].__args__[0]
+        else:
+          field_class = getattr(self, "__annotations__")[field_name]
+        # Read field_val from dict
+        field_val = field_class()._from_dict(  # pylint: disable=protected-access
```

Comment:
here field_class can be typing.Union which will raise Error when call field_class()

```python
# FIX Optional[T], add code after line 73 will fix it
if hasattr(field_class, "__args__"):
    field_class = field_class.__args__[0]
```